apps/aecore/test/aec_block_genesis_tests.erl

```diff
@@ -0,0 +1,113 @@
+-module(aec_block_genesis_tests).
+
+-include_lib("eunit/include/eunit.hrl").
+-include("blocks.hrl").
+
+-define(TEST_MODULE, aec_block_genesis).
+
+%%%===================================================================
+%%% Test cases
+%%%===================================================================
+
+genesis_block_test_() ->
+    {foreach,
+     fun() ->
+         meck:new(aec_genesis_block_settings, [passthrough]),
+         ok
+     end,
+     fun(ok) ->
+         meck:unload(aec_genesis_block_settings),
+         ok
+     end,
+     [ {"Genesis block is commutative according to the list of preset accounts",
+        fun() ->
+            [Account1, Account2, Account3] = generate_accounts(3),
+
+            meck_preset_accounts([Account1, Account2, Account3]),
+            Header1 = ?TEST_MODULE:genesis_header(),
+            {ok, Header1Hash} = aec_headers:hash_header(Header1),
+
+            % same accounts shuffled
+            meck_preset_accounts([Account3, Account1, Account2]),
+            Header2 = ?TEST_MODULE:genesis_header(),
+            {ok, Header2Hash} = aec_headers:hash_header(Header2),
+
+            % produced genesis headers are the same
+            ?assertEqual(Header1, Header2),
+            ?assertEqual(Header1Hash, Header2Hash),
+            ok
+        end},
+       {"Genesis block is idempotent according to the list of preset accounts",
+        fun() ->
+            [Account1, Account2, Account3] = generate_accounts(3),
+
+            meck_preset_accounts([Account1, Account2, Account3]),
+            Header1 = ?TEST_MODULE:genesis_header(),
+            {ok, Header1Hash} = aec_headers:hash_header(Header1),
+
+            % same accounts, one is present twice
+            meck_preset_accounts([Account1, Account2, Account3, Account1]),
+            Header2 = ?TEST_MODULE:genesis_header(),
+            {ok, Header2Hash} = aec_headers:hash_header(Header2),
+
+            % produced genesis headers are the same
+            ?assertEqual(Header1, Header2),
+            ?assertEqual(Header1Hash, Header2Hash),
+            ok
+        end}]}.
+
+preset_trees_test_() ->
+    {foreach,
+     fun() ->
+         meck:new(aec_genesis_block_settings, [passthrough]),
+         ok
+     end,
+     fun(ok) ->
+         meck:unload(aec_genesis_block_settings),
+         ok
+     end,
+     [ {"Trees contain expected accounts",
+        fun() ->
+            PresetAccounts = generate_accounts(10),
+
+            meck_preset_accounts(PresetAccounts),
+            Trees = ?TEST_MODULE:populated_trees(),
+            Accounts = aec_trees:accounts(Trees),
+            lists:foreach(
+                fun({Pubkey, Balance}) ->
+                    Acc = aec_accounts_trees:get(Pubkey, Accounts),
+                    ActualBalance = aec_accounts:balance(Acc),
+                    ?assertEqual(Balance, ActualBalance)
+                end,
+                PresetAccounts),
```

Comment:
Additional check could be that no account further to the once preset are present in the accounts tree. Completely optional.

Reply:
Addressed in [5ee23a](https://github.com/aeternity/epoch/pull/1745/commits/5ee23a9c7082d93b8b931941efb50582f58b8e95)